HPCC-27887 Implement new keyed compression format

ghalliday opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman commented on system/jhtree/ctfile.hpp:
```diff
@@ -92,11 +98,12 @@ struct __declspec(novtable) jhtree_decl KeyHdr
     __int64 bloomHead; /* fpos of bloom table data, if present 100x */
-    __uint64 partitionFieldMask; /* Bitmap indicating partition keyed fields */
+    __uint64 partitionFieldMask; /* Bitmap indicating partition keyed fields 108x */
+    unsigned getFlags() { return keyFlags; }
```
Won't this fail on an old index? Do you need to fall back to returning oldktype if keyFlags is (whatever value we can be sure is in old index files).

richardkchapman commented on system/jhtree/ctfile.hpp:
```diff
@@ -161,14 +185,15 @@ class jhtree_decl CKeyHdr : public CWritableKeyNode
     inline unsigned getNodeSize() { return hdr.nodeSize; }
     inline bool hasSpecialFileposition() const { return true; }
-    inline bool isRowCompressed() const { return (hdr.ktype & (HTREE_QUICK_COMPRESSED_KEY|HTREE_VARSIZE)) == HTREE_QUICK_COMPRESSED_KEY; }
+    inline bool isRowCompressed() const { return (hdr.keyFlags & (HTREE_QUICK_COMPRESSED_KEY|HTREE_VARSIZE)) == HTREE_QUICK_COMPRESSED_KEY; }
```
Similar compatibility issues...

richardkchapman reviewed (changes requested):
Some initial concerns from first (very quick) scan

richardkchapman commented on system/jhtree/keybuild.cpp:
```diff
@@ -168,11 +170,24 @@ class CKeyBuilder : public CInterfaceOf<IKeyBuilder>
                 }
             }
+
+            //Create a representation of the null row - used for the new compression algorithm
```
Should this be conditional?

ghalliday committed:
HPCC-27888 Implement new keyed compression format

Various changes to the structure of jhtree classes to allow future
expansion for new node types and compression options.

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

richardkchapman commented on system/jhtree/keybuild.cpp:
```diff
@@ -119,6 +114,7 @@ class CKeyBuilder : public CInterfaceOf<IKeyBuilder>
     CRC32EndHT crcEndPosTable;
     CRC32 headCRC;
     bool doCrc = false;
+    KeyBuildContext ctx;
```
Is this used?

ghalliday replied:
no, historic location prior to refactoring

richardkchapman commented on system/jhtree/jhinplace.hpp:
```diff
@@ -0,0 +1,252 @@
+    virtual CWriteNode *createNode(offset_t _fpos, CKeyHdr *_keyHdr, bool isLeafNode) const override
+    {
+        if (isLeafNode)
+            return new CInplaceLeafWriteNode(_fpos, _keyHdr, const_cast<KeyBuildContext &>(ctx));
```
cast seems unnecessary/confusing

ghalliday replied:
changed the member so it is mutable - much cleaner

richardkchapman commented on system/jhtree/jhinplace.cpp:
```diff
@@ -0,0 +1,2681 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
Wrong date

richardkchapman commented on system/jhtree/jhinplace.cpp:
```diff
@@ -0,0 +1,2681 @@
+static constexpr byte specialSingleChars[31] = {
+    0, 1, 2, 3, 4, 5, 6, 7,                 // 1-8      Useful for small numbers
+    ' ', '0', 'F', 'M', 'U',                // 9-13     Single padding and gender
+                                            // 14-31    Currently unused.
```
Other digits would seem likely to occur?

ghalliday replied:
This is something that would be good to tune on some real-life indexes.

richardkchapman commented on system/jhtree/jhinplace.cpp:
```diff
@@ -0,0 +1,2681 @@
+            clean.append('.');
+    }
+
+    printf("%*s(%s[%s] %u:%u[%u] [%s]", indent, "", clean.str(), dataHex.str(), data.length(), squashedData.length(), getSize(), squashedText.str());
```
DBGLOG would be more normal...

richardkchapman commented on system/jhtree/jhinplace.cpp:
```diff
@@ -0,0 +1,2681 @@
+
+void PartialMatchBuilder::add(size32_t len, const void * data)
+{
+    //It is legal to add rows longer than keyLen, but cannot strip trailing nulls
```
Why would you want to?

ghalliday replied:
Previous iteration of this code where it was also being used for compressing the payload - will remove the comment.

richardkchapman commented on system/jhtree/jhinplace.cpp:
```diff
@@ -0,0 +1,2681 @@
+    // Locate first record greater than src
+    while ((int)a<b)
+    {
+        //MORE: Note sure why the index is subtly different to the GTE version
```
typo: Note -> not

richardkchapman commented on system/jhtree/jhinplace.cpp:
```diff
@@ -0,0 +1,2681 @@
+
+        if (isLeaf())
+        {
+            //Extra position serialized for leafs..
```
leaves? leaf nodes?

richardkchapman commented on system/jhtree/jhinplace.cpp:
```diff
@@ -0,0 +1,2681 @@
+
+InplaceIndexCompressor::~InplaceIndexCompressor()
+{
+    delete [] ctx.nullRow;
```
Shouldn't ctc destructor be responsible for this?

ghalliday replied:
moved

richardkchapman reviewed (commented):
Some initial comments from a quick scan

richardkchapman reviewed (approved):
Changes in second commit look good. Worth squashing and merging I think

richardkchapman commented:
Ditto third commit
